pkg/controller/direct/compute/networkattachment_controller.go

```diff
@@ -0,0 +1,325 @@
+	}
+	networkAttachmentClient, err := gcpClient.newNetworkAttachmentsClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &NetworkAttachmentAdapter{
+		gcpClient: networkAttachmentClient,
+		id:        id,
+		desired:   obj,
+	}, nil
+}
+
+func (m *networkAttachmentModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type NetworkAttachmentAdapter struct {
+	gcpClient *compute.NetworkAttachmentsClient
+	id        *v1alpha1.NetworkAttachmentIdentity
+	desired   *krm.ComputeNetworkAttachment
+	actual    *computepb.NetworkAttachment
+}
+
+var _ directbase.Adapter = &NetworkAttachmentAdapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return false means the object is not found. This triggers Adapter `Create` call.
+// Return a non-nil error requeues the requests.
+func (a *NetworkAttachmentAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting NetworkAttachment", "name", a.id)
+
+	req := &computepb.GetNetworkAttachmentRequest{
+		Project:           a.id.Parent().ProjectID,
+		Region:            a.id.Parent().Location,
+		NetworkAttachment: a.id.ID(),
+	}
+	actual, err := a.gcpClient.Get(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting NetworkAttachment %q: %w", a.id, err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *NetworkAttachmentAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating NetworkAttachment", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := ComputeNetworkAttachmentSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	resource.Name = direct.LazyPtr(a.id.ID())
+
+	req := &computepb.InsertNetworkAttachmentRequest{
+		Project:                   a.id.Parent().ProjectID,
+		Region:                    a.id.Parent().Location,
+		NetworkAttachmentResource: resource,
+	}
+	op, err := a.gcpClient.Insert(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating NetworkAttachment %s: %w", a.id, err)
+	}
+
+	err = op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("compute NetworkAttachment %s waiting creation: %w", a.id.String(), err)
+	}
+	log.Info("successfully created compute NetworkAttachment in gcp", "name", a.id)
+
+	// Get the created resource
+	created := &computepb.NetworkAttachment{}
+	getReq := &computepb.GetNetworkAttachmentRequest{
+		Project:           a.id.Parent().ProjectID,
+		Region:            a.id.Parent().Location,
+		NetworkAttachment: a.id.ID(),
+	}
+	created, err = a.gcpClient.Get(ctx, getReq)
```

Comment:
Same here, maybe put the shared "get" logic into a separate function so that they can be reused by Find, Create and Update?